Building without default features doesn't work

I tried building the image example without any default features and I got compile errors.

The following seem necessary to get a simple compile working: fonts, saving, sounds. It looks like `FontError`, `SoundError`, and `SerdeError` are used in `error.rs` and the build fails when they're not defined.

There should be a CI test for building without any features defined to catch stuff like this.

If my experiments with this library work out, I might be able to contribute a fix for this.

Sure:

```
error[E0201]: duplicate definitions with name `provide_gamepads_impl`:
   --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/input/gamepad.rs:132:5
    |
107 | /     fn provide_gamepads_impl(&self, buffer: &mut Vec<Gamepad>) {
108 | |         fn new(id: u32) -> Gamepad {
109 | |             Gamepad { id, buttons: [ButtonState::NotPressed; 17], axes: [0.0; 4] }
110 | |         }
...   |
128 | |         }
129 | |     }
    | |_____- previous definition of `provide_gamepads_impl` here
...
132 | /     fn provide_gamepads_impl(&self, _buffer: &mut Vec<Gamepad>) {
133 | |         //Inentionally a no-op
134 | |     }
    | |_____^ duplicate definition

error: aborting due to previous error

For more information about this error, try `rustc --explain E0201`.
error: Could not compile `quicksilver`.
```

And my `Cargo.toml`:

    quicksilver = { version = "0.2.1", default-features = false, features = ["fonts", "saving", "sounds"]  }

If I disable all default features, desktop builds don't work:

```
error[E0599]: no variant named `SoundError` found for type `error::QuicksilverError` in the current scope
  --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/error.rs:46:14
   |
20 | pub enum QuicksilverError {
   | ------------------------- variant `SoundError` not found here
...
46 |             &QuicksilverError::SoundError(ref err) => err.description(),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ variant not found in `error::QuicksilverError`

error[E0599]: no variant named `SerdeError` found for type `error::QuicksilverError` in the current scope
  --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/error.rs:48:14
   |
20 | pub enum QuicksilverError {
   | ------------------------- variant `SerdeError` not found here
...
48 |             &QuicksilverError::SerdeError(ref err) => err.description(),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ variant not found in `error::QuicksilverError`

error[E0599]: no variant named `FontError` found for type `error::QuicksilverError` in the current scope
  --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/error.rs:49:14
   |
20 | pub enum QuicksilverError {
   | ------------------------- variant `FontError` not found here
...
49 |             &QuicksilverError::FontError(ref err) => err.description()
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ variant not found in `error::QuicksilverError`

error[E0599]: no variant named `SoundError` found for type `error::QuicksilverError` in the current scope
  --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/error.rs:57:14
   |
20 | pub enum QuicksilverError {
   | ------------------------- variant `SoundError` not found here
...
57 |             &QuicksilverError::SoundError(ref err) => Some(err),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ variant not found in `error::QuicksilverError`

error[E0599]: no variant named `SerdeError` found for type `error::QuicksilverError` in the current scope
  --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/error.rs:59:14
   |
20 | pub enum QuicksilverError {
   | ------------------------- variant `SerdeError` not found here
...
59 |             &QuicksilverError::SerdeError(ref err) => Some(err),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ variant not found in `error::QuicksilverError`

error[E0599]: no variant named `FontError` found for type `error::QuicksilverError` in the current scope
  --> /home/otto/.cargo/registry/src/github.com-1ecc6299db9ec823/quicksilver-0.2.1/src/error.rs:60:14
   |
20 | pub enum QuicksilverError {
   | ------------------------- variant `FontError` not found here
...
60 |             &QuicksilverError::FontError(ref err) => Some(err)
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ variant not found in `error::QuicksilverError`

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0599`.
error: Could not compile `quicksilver`.

To learn more, run the command again with --verbose.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -19,10 +19,19 @@ rust:
 matrix:
   include:
     - rust: nightly
+      env: WASM=0
+      script:
+        - cargo check --examples --no-default-features
+    - rust: nightly
       env: WASM=1
       install: rustup target add wasm32-unknown-unknown
       script:
         - cargo +nightly check --target wasm32-unknown-unknown --examples
+    - rust: nightly
+      env: WASM=1
+      install: rustup target add wasm32-unknown-unknown
+      script:
+        - cargo +nightly check --target wasm32-unknown-unknown --examples --no-default-features
   allow_failures:
     - rust: nightly
       env: WASM=0

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,7 +11,7 @@ readme = "README.md"
 repository = "https://github.com/ryanisaacg/quicksilver"
 
 [features]
-default = ["collisions", "fonts", "gamepads", "complex_shapes", "immi_ui", "saving", "sounds"]
+default = ["collisions", "fonts", "gamepads", "saving", "sounds"]
 
 collisions = ["nalgebra", "ncollide2d"]
 complex_shapes = ["lyon"]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -80,7 +80,6 @@
 extern crate futures;
 extern crate image;
 extern crate rand;
-extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 
@@ -109,6 +108,8 @@ extern crate ncollide2d;
 extern crate rodio;
 #[cfg(feature = "rusttype")]
 extern crate rusttype;
+#[cfg(feature = "serde_json")]
+extern crate serde;
 #[cfg(feature = "serde_json")]
 extern crate serde_json;
 

--- a/src/lifecycle/gamepad_provider.rs
+++ b/src/lifecycle/gamepad_provider.rs
@@ -99,7 +99,7 @@ impl GamepadProvider {
             }));
     }
 
-    #[cfg(any(target_os="macos", not(feature = "gamepads")))]
+    #[cfg(all(not(target_arch = "wasm32"), any(target_os="macos", not(feature = "gamepads"))))]
     fn provide_gamepads_impl(&self, _buffer: &mut Vec<Gamepad>) {
         //Inentionally a no-op
     }

